rust-runtime/smithy-http/src/event_stream.rs

```diff
@@ -181,23 +183,35 @@ impl<T, E> Receiver<T, E> {
         }
     }
 
+    async fn buffer_next_chunk(&mut self) -> Result<(), SdkError<E, Message>> {
+        if !self.stream_ended {
+            let next_chunk = self
+                .body
+                .data()
+                .await
+                .transpose()
+                .map_err(|err| SdkError::DispatchFailure(err))?;
+            if let Some(chunk) = next_chunk {
+                // The SegmentedBuf will automatically purge when it reads off the end of a chunk boundary
+                self.buffer.push(chunk);
+            } else {
+                self.stream_ended = true;
+            }
+        }
+        Ok(())
+    }
+
     async fn next_message(&mut self) -> Result<Option<Message>, SdkError<E, Message>> {
-        let next_chunk = self
-            .body
-            .data()
-            .await
-            .transpose()
-            .map_err(|err| SdkError::DispatchFailure(err))?;
-        if let Some(chunk) = next_chunk {
-            // The SegmentedBuf will automatically purge when it reads off the end of a chunk boundary
-            self.buffer.push(chunk);
+        while !self.stream_ended {
```

Comment:
I _think_ this is correct, but it's pretty subtle and relies on `buffer_next_chunk` never being called without first trying to decode a frame from the buffer. 

Otherwise, buffer next chunk would set `stream_ended` to true, and we'd never enter this loop, even though there was still buffered data.

I think I'd prefer an additional check that `self.buffer` is empty before we bail out of this loop.